### PR TITLE
egctl: add `version` command

### PR DESCRIPTION
**Pull request by zirain**
xref: https://github.com/envoyproxy/gateway/issues/915

### CODE REVIEW

**Comment by Xunzhuo** [comment]
Can you sparate the design docs into issues in case we can better track it or make contributors/maintainers can help implement it? https://github.com/envoyproxy/gateway/blob/main/docs/latest/design/egctl.md @zirain

File: internal/ctl/app/version.go
Comment (danehans) [inline internal/ctl/app/version.go:41]: Since this is user-facing, I prefer yaml as the default over json.

```
$ egctl version --remote -o yaml
client: v0.2.0
servers:
  envoy-gateway-system/envoy-gateway-84f8b94f-8tqq7:
    envoyGatewayVersion: v0.2.0
    envoyVersion: latest
    gatewayAPIVersion: v0.6.0
    gitCommitID: 6f8b6fbeab204f9d5b334fb1e1cb6fbdc8b33112
```

Can you also update the following to the version output:
- Have the version command return `server` instead of `servers.
- Update `envoyVersion` to `envoyProxyVersion`.
- Parse `envoy-gateway-system/envoy-gateway-84f8b94f-8tqq7` to provide additional context in the output. For example:
```
namespace: envoy-gateway-system
deploymentName: envoy-gateway
```

File: internal/ctl/app/version.go
Comment (danehans) [inline internal/ctl/app/version.go:77]: The EG namespace should be exposed as a flag that defaults to `envoy-gateway`.

File: internal/ctl/app/version.go
Comment (danehans) [inline internal/ctl/app/version.go:48]: `ServerVersions` should be updated if the `remote` flag is removed and `egctl version` always returns the server version.

File: internal/ctl/app/version.go
Comment (danehans) [inline internal/ctl/app/version.go:63]: s/build CLI client fail/failed to build Kubernetes client/

File: cmd/egctl/main.go
Comment (danehans) [reply cmd/egctl/main.go:14]: Is the `app` directory for `egctl` being used to separate `egctl` and `envoy-gateway` builds? Would it be better to reorganize and use the current `internal/cmd` structure? For example:
```
internal/cmd/version
internal/cmd/envoy-gateway
internal/cmd/egctl
```

File: internal/ctl/app/version.go
Comment (danehans) [inline internal/ctl/app/version.go:79]: s/fail/failed/

Other parts of the code use `%s/%s` with nn.Namespace and nn.Name.

File: internal/ctl/app/version.go
Comment (danehans) [inline internal/ctl/app/version.go:84]: Other parts of the code use `%s/%s` with nn.Namespace and nn.Name.

File: internal/utils/kube/client.go
Comment (danehans) [inline internal/utils/kube/client.go:39]: Why not use the existing `GetScheme()`?

File: internal/utils/kube/client.go
Comment (danehans) [inline internal/utils/kube/client.go:1]: IMO this should be part of a new pkg `internal/kubernetes`.

File: internal/utils/kube/client.go
Comment (danehans) [inline internal/utils/kube/client.go:25]: Why not use client-go?

**Review by danehans — CHANGES_REQUESTED** [review]

File: internal/ctl/app/version.go
Comment (zirain) [reply internal/ctl/app/version.go:30]: same as https://github.com/envoyproxy/gateway/blob/2479b9391055d4b0958e8f25b9966c968d8d75a8/internal/cmd/versions.go#L19

should we change both?

File: internal/ctl/app/version.go
Comment (zirain) [reply internal/ctl/app/version.go:77]: here is not namespace, it's container name.
BTW, it should support customization.

File: internal/utils/kube/client.go
Comment (zirain) [reply internal/utils/kube/client.go]: I don't understand your point.

File: internal/ctl/app/version.go
Comment (zirain) [reply internal/ctl/app/version.go]: /hold https://github.com/envoyproxy/gateway/pull/956

File: internal/cmd/egctl/version.go
Comment (danehans) [inline internal/cmd/egctl/version.go:37]: /versions/version/

File: cmd/egctl/main.go
Comment (danehans) [inline cmd/egctl/main.go:32]: s/NewVersionsCommand/NewVersionCommand/

File: internal/cmd/egctl/version.go
Comment (danehans) [inline internal/cmd/egctl/version.go:28]: s/NewVersionsCommand/NewVersionCommand/

File: internal/cmd/egctl/version.go
Comment (danehans) [inline internal/cmd/egctl/version.go:47]: I don't expect the EG container name to change so I'm not sure why the container name needs to be exposed to a user. However, I can see EG being run in a different ns than `envoy-gateway-system` and expose the ns to the user while defaulting to "envoy-gateway-system".

**Review by danehans — CHANGES_REQUESTED** [review]

File: internal/cmd/egctl/version.go
Comment (zirain) [reply internal/cmd/egctl/version.go:47]: `version` will search all namespace, expect vendors may use EG multi-tenancy with custom container name？

**Comment by zirain** [comment]
output of running `make build && bin/linux/amd64/egctl v`

```
client: v0.2.0
server:
- Name: envoy-gateway-6cbd765bdc-ddqpm
  Namespace: envoy-gateway-system
  envoyGatewayVersion: v0.2.0
  envoyProxyVersion: latest
  gatewayAPIVersion: v0.6.0
  gitCommitID: 54fdd5bc185d8fa57e7afa7ec5f53bb1466757e0
```

File: internal/cmd/egctl/version.go
Comment (danehans) [reply internal/cmd/egctl/version.go:47]: I'm not following your point here about the EG container name. For multitenancy, a user would install EG in different namespaces with each EG using a different `controllerName`. A user would then create a GatewayClass per EG and one or more Gateways that reference the GatewayClass. By default, egctl should expect EG in the `envoy-gateway-system` namespace with the Deployment name of `envoy-gateway`. If egctl needs to communicate with an EG in a different namespace, a flag such as `--namespace` should be surfaced. For security purposes, we should avoid providing egctl access across all namespaces.

**Comment by danehans** [comment]
I just tested egctl and it's showing the incorrect client version. Maybe this PR needs to be rebased?
```
$ ./bin/darwin/amd64/egctl version
client: v0.2.0
server:
- Name: envoy-gateway-7fd4c7b9cf-jpjrt
  Namespace: envoy-gateway-system
  envoyGatewayVersion: v0.3.0
  envoyProxyVersion: v1.25-latest
  gatewayAPIVersion: v0.6.1
  gitCommitID: f735041d6ce31e104bf00f9a96f47d327634a923
```

In the future, EG could be running multiple replicas. In this case, which pod name will be displayed? It might be better to surface the deployment name instead.

File: internal/cmd/egctl/version.go
Comment (zirain) [reply internal/cmd/egctl/version.go:47]: let me remove this part first.

**Comment by zirain** [comment]
> I just tested egctl and it's showing the incorrect client version. Maybe this PR needs to be rebased?
> 
> ```
> $ ./bin/darwin/amd64/egctl version
> client: v0.2.0
> server:
> - Name: envoy-gateway-7fd4c7b9cf-jpjrt
>   Namespace: envoy-gateway-system
>   envoyGatewayVersion: v0.3.0
>   envoyProxyVersion: v1.25-latest
>   gatewayAPIVersion: v0.6.1
>   gitCommitID: f735041d6ce31e104bf00f9a96f47d327634a923
> ```
> 
> In the future, EG could be running multiple replicas. In this case, which pod name will be displayed? It might be better to surface the deployment name instead.

after rebasing from main, it's correct now.

**Comment by danehans** [comment]
@zirain thanks for rebasing. I have confirmed that the client/server versions now match. As a follow-on to https://github.com/envoyproxy/gateway/pull/914#issuecomment-1426105378, EG can run multiple replicas in the future. In this case, which pod name will be displayed? I think it's better to surface the EG deployment name instead of the pod name. Thoughts?

**Comment by zirain** [comment]
> In the future, EG could be running multiple replicas. In this case, which pod name will be displayed? It might be better to surface the deployment name instead.

for now, envoy pod will displayed.

**Comment by danehans** [comment]
Why display the Pod name instead of the Deployment name?

**Comment by zirain** [comment]
> Why display the Pod name instead of the Deployment name?

if you look at the implement, `egctl` list pods by selector, then gather server information via `PodExec`.
BTW, I have no strong option about change to display deployment name.

**Comment by danehans** [comment]
Yes, I understand the implementation. I just don't understand the purpose of displaying the EG pod name instead of the deployment name. I'm fine with resolving this in the future when EG supports multiple replicas.

**Comment by danehans** [comment]
@zirain I can approve after you resolve the merge conflicts.

**Review by danehans — APPROVED** [review]

**Comment by Xunzhuo** [comment]
> Yes, I understand the implementation. I just don't understand the purpose of displaying the EG pod name instead of the deployment name. I'm fine with resolving this in the future when EG supports multiple replicas.

I prefer deployment name too, but let us get this in first. We can make it later.

**Review by Xunzhuo — APPROVED** [review]
Thanks @zirain !

**Pull request merged**
